jaiarobotics/jaiabot
# bug/task-packet-date-selection-start-date

# bug/task-packet-date-selection-start-date

Note: `date.toISOString()` converts the date object to UTC taking into account the timezone of the date object

**TaskPacket Start:** 
```
11/12/2023, 14:28:17
```

**Client:**
```
// Console output from client after setting start date/time to Nov 12 2023 14:28
// Local time converted to UTC
date (preAdj): Sun Nov 12 2023 14:28:00 GMT-0500 (Eastern Standard Time)
date (postISOAdj): 2023-11-12 19:28:00
startDateStr: 2023-11-12 19:28:00 // Value passed to server
```

**Server:**
```
# Print statements
logging.warning(f'{start_date}') # Value from client
logging.warning(f'[{utime(start_date)}]')
# Only single test drift TaskPacket from above is in task_packets_list
logging.warning(
    list(map(
        lambda task_packet: int(task_packet['start_time']),task_packets_list
    ))
)

# Print statements output
WARNING:root:2023-11-12 19:28:00
WARNING:root:[1699817280000000] # Timestamp from client
WARNING:root:[1699817297000000] # Timestamp from test drift TaskPacket

```

**Explanation:**
```
* Our test drift TaskPacket registered at 14:28:17 local time (19:28:17 UTC)

* The TaskPacket start date set by the operator in the Edit Dates 
  view comes in as local time
* That local time is then converted to UTC via the method toISOString()

* Now the time passed to the server from the client and the TaskPacket time on
  the server are both in UTC
* This is confirmed by the server output
  Note the gap of 17 seconds from the first timestamp to the second one
  The first represents the value from the client...
  The operator can only set time to the nearest minute which is why
  they could not choose the exact start time of the TaskPacket,
  however, the 17 second gap shows both the client and server are in UTC

```

## Tests:
1) Play a two waypoint run with waypoint 1 assigned to a dive and waypoint 2 assigned to a drift
    * Hide the drift from waypoint 2 and only show the dive (and its associated drift) by adjusting the end date/time
        * PASS
    * Undo this change by adjusting the date/time
        * FAIL -- I checked the Keep End Date Current checkbox, the time updated in the calendar, but the drift packet did not reappear. When the checkbox goes from unchecked to checked a new request should be sent to the server
        * PASS (after code changes)
    * Hide the dive packet from waypoint 1 and only show the drfit from waypoint 2 by changing the start and end date/times
        * PASS
    * Undo this change by adjusting the date/time
        * FAIL -- I toggled the Edit Dates off...the dive packet reappeared but when I opened the Edit Dates view again the start time did not reset on the calendar
        * PASS (after code changes)

2) With TaskPackets loaded from June in the offload directory, make them appear along with the newly created TaskPackets
    * Make the old TaskPackets appear without showing the new ones
        * PASS
    * Make the old TaskPackets disapear and show the new ones
        * FAIL -- The old TaskPackets would not disapear...I think the server was being overwhelmed by the number of requests becasue there were two request calls for TaskPackets within a for loop for each TaskPacket
        * PASS (after code changes)

3) Line tool test...remove TaskPackets down the line (one by one), then add them back one by one
    * With 1 bot, 3 dive tasks
        * PASS
    * With 2 bots, 3 dive tasks each
        * PASS
    * These test failed on Matt's computer, so I just pushed changes with debug statements
